dev/NumberBox/NumberBox.cpp

```diff
@@ -0,0 +1,387 @@
+        {
+            Value(Maximum());
+        }
+        else
+        {
+            Value(Minimum());
+        }
+    }
+}
+
+void NumberBox::ValidateInput()
+{
+    // Validate the content of the inner textbox
+    if (auto textBox = m_textBox.get())
+    {
+        auto text = textBox.Text();
+        
+        // Handles empty TextBox case, set text to current value
+        if (text.empty())
+        {
+            UpdateTextToValue();
+        }
+        else
+        {
+            if (AcceptsCalculation() && IsFormulaic(text))
+            {
+                NormalizeShorthandOperations();
+                EvaluateInputCalculation();
+            }
+
+            // Setting NumberFormatter to something that isn't an INumberParser will throw an exception, so this should be safe
+            auto numberParser = NumberFormatter().as<winrt::INumberParser>();
+            auto parsedNum = numberParser.ParseDouble(text);
+
+            if (!parsedNum)
+            {
+                if (BasicValidationMode() == winrt::NumberBoxBasicValidationMode::InvalidInputOverwritten)
+                {
+                    // Override text to current value
+                    UpdateTextToValue();
+                }
+            }
+            else
+            {
+                Value(parsedNum.Value());
+            }
+        }
+    }
+}
+
+void NumberBox::OnSpinDownClick(winrt::IInspectable const&  sender, winrt::RoutedEventArgs const& args)
+{
+    StepValue(false);
+}
+
+void NumberBox::OnSpinUpClick(winrt::IInspectable const& sender, winrt::RoutedEventArgs const& args)
+{
+    StepValue(true);
+}
+
+void NumberBox::OnNumberBoxKeyUp(winrt::IInspectable const& sender, winrt::KeyRoutedEventArgs const& args)
+{
+    switch (args.OriginalKey())
+    {
+        case winrt::VirtualKey::Enter:
+        case winrt::VirtualKey::GamepadA:
+            ValidateInput();
+            break;
+
+        case winrt::VirtualKey::Escape:
+        case winrt::VirtualKey::GamepadB:
+            UpdateTextToValue();
+            break;
+
+        case winrt::VirtualKey::Up:
+            StepValue(true);
+            break;
+
+        case winrt::VirtualKey::Down:
+            StepValue(false);
+            break;
+    }
+}
+
+void NumberBox::OnScroll(winrt::IInspectable const& sender, winrt::PointerRoutedEventArgs const& args)
+{
+    if (HyperScrollEnabled())
+    {
+        int delta = args.GetCurrentPoint(*this).Properties().MouseWheelDelta();
+        if (delta > 0)
+        {
+            StepValue(true);
+        }
+        else if (delta < 0)
+        {
+            StepValue(false);
+        }
+    }
+}
+
+void NumberBox::StepValue(bool isPositive)
+{
+    double oldVal = Value();
+    double newVal = oldVal;
+
+    if (isPositive)
+    {
+        newVal += StepFrequency();
+    }
+    else
+    {
+        newVal -= StepFrequency();
+    }
+
+    if (WrapEnabled())
+    {
+        if (newVal > Maximum())
+        {
+            newVal = Minimum();
```

Comment:
>newVal = Minimum(); [](start = 12, length = 19)

hmm, we don't consider how far over the max the value was when we wrap? For instance if i have a min and max of 0 and 100, my current value is 98 and my step frequency is 5 and i hit the up button, I think i would expect the value to be 2, not 0.

Reply:
I actually did that intentionally because I thought the opposite! I'll add it as a spec question since it's not explained.